### PR TITLE
nsq: 0.3.5 -> 1.1.0, add nixos/nsq

### PR DESCRIPTION
**Pull request by pjones**
###### Motivation for this change

Upgrade nsq to the latest release and create a NixOS module for it.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Infinisil** [comment]
Take a look at NixOS/rfcs#42. With the approach described there, the implementation of this will be simplified and also get correct merging behavior (`recursiveUpdate` doesn't merge correctly).

File: nixos/modules/misc/ids.nix [hunk 1]
```diff
@@ -339,6 +339,7 @@
       rss2email = 312;
       cockroachdb = 313;
       zoneminder = 314;
+      nsq = 315;
```
Comment (flokli) [inline nixos/modules/misc/ids.nix]: I guess we can avoid allocating a user, by using the `DynamicUser` feature of systemd.

File: nixos/modules/services/networking/nsq.nix [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    } ''
+      remarshal -if json -of toml \
+        < ${pkgs.writeText "${name}.json" (builtins.toJSON attrs)} \
+        > $out
+    '';
+
+  # Configuration for nsqd as a Nix attrset:
+  nsqdConfig = recursiveUpdate {
+    mem_queue_size = cfg.nsqd.queueSize;
+    data_path = "${cfg.dataDir}/nsqd";
```
Comment (flokli) [inline nixos/modules/services/networking/nsq.nix]: This would be `/var/lib/nsqd`.

File: nixos/modules/services/networking/nsq.nix [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  # Configuration for nsqadmin written into the Nix store as TOML:
+  nsqadminConfigFile = toTOML "nsqadmin" nsqadminConfig;
+
+  # Generate a systemd service:
+  mkService = name: configFile: {
+    description = "${name} NSQ daemon";
+    wantedBy = [ "multi-user.target" ];
+    serviceConfig.User = cfg.user;
+    serviceConfig.Group = cfg.group;
+    serviceConfig.PermissionsStartOnly = true;
```
Comment (flokli) [inline nixos/modules/services/networking/nsq.nix]: we can drop this, and let systemd handle creation of the state directory.

File: nixos/modules/misc/ids.nix (hunk 1 above)
Comment (Infinisil) [reply nixos/modules/misc/ids.nix]: And if `DynamicUser` doesn't work, just a normal user without a static uid will work too

**Comment by pjones** [comment]
Turns out that I'm not going to need NSQ in NixOS right now.  I'll revisit this in the future.

**Pull request closed without merging**